posts/2018-09-04-do-not-use-list-monoid.md

```diff
@@ -0,0 +1,66 @@
+Similar to `List` it is also a very simple data structure, but unlike `List` it supports both constant time `append` and `prepend`.
+This makes it's `Monoid` instance super performant and a much better fit for usage with `Validated`,`Writer`, `Ior` or `Const`.
+
+To utilize this, we've added a bunch of shorthands in Cats 1.3 that previously used `NonEmptyList` to use `NonEmptyChain` instead. These include type aliases like `ValidatedNec` or `IorNec` as well as helper functions like `groupByNec` or `Validated.invalidNec`.
+We hope that these make it easy for you to upgrade to the more efficient data structure and enjoy those benefits as soon as possible.
+
+To get a good idea of the performance improvements, here are some benchmarks that test monoidal append to convince you of its efficiency.
+
+```
+[info] Benchmark                                  Mode  Cnt   Score   Error  Units
+[info] CollectionMonoidBench.accumulateChain     thrpt   20  51.911 ± 7.453  ops/s
+[info] CollectionMonoidBench.accumulateList      thrpt   20   6.973 ± 0.781  ops/s
+[info] CollectionMonoidBench.accumulateVector    thrpt   20   6.304 ± 0.129  ops/s
+```
+
+As you can see accumulating things with `Chain` is more than 7 times faster than `List` and over 8 times faster than `Vector`.
+So appending seems to be a lot more performant than the standard library collections, but what about operations like `map` or `fold`.
+Fortunately we've also benchmarked these:
+
+```
+[info] Benchmark                           Mode  Cnt          Score         Error  Units
+[info] ChainBench.foldLeftLargeChain      thrpt   20        117.267 ±       1.815  ops/s
+[info] ChainBench.foldLeftLargeList       thrpt   20        135.954 ±       3.340  ops/s
+[info] ChainBench.foldLeftLargeVector     thrpt   20         61.613 ±       1.326  ops/s
+[info]
+[info] ChainBench.mapLargeChain           thrpt   20         59.379 ±       0.866  ops/s
+[info] ChainBench.mapLargeList            thrpt   20         66.729 ±       7.165  ops/s
+[info] ChainBench.mapLargeVector          thrpt   20         61.374 ±       2.004  ops/s
+```
+
+While not as dominant, `Chain` can still hold its ground fairly well. 
+It won't have the random access performance of something like `Vector`, but in almost all other cases, `Chain` seems to outperform it quite handily.
+So if you don't perform a lot of random access on your data structure (I certainly don't), then you should be fine using `Chain` extensively instead.
+
+So next time you write any code that uses `List` or `Vector` as a `Monoid`, be sure to use `Chain` instead!
+
+The whole code for `Chain` and `NonEmptyChain` can be found [here](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/data/Chain.scala) and [here](https://github.com/typelevel/cats/blob/master/core/src/main/scala/cats/data/NonEmptyChain.scala).
```

Comment:
Use permalinks (e.g. with the tag) here.

Reply:
I'm not sure what you mean by this

Reply:
These links:
https://github.com/typelevel/cats/blob/v1.3.0/core/src/main/scala/cats/data/Chain.scala
https://github.com/typelevel/cats/blob/v1.3.0/core/src/main/scala/cats/data/NonEmptyChain.scala

Reply:
(As opposed to links to `master`, these ones will still work even after someone moves around files.)

Reply:
Right, of course, sorry and thanks :smile: